### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Benawi**
### General requirements
- Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
- Make sure that you used the correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
- Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
### Ruby requirements
- Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).
### Project requirements
- [x] All methods visible in the [diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be implemented
- [x] All properties visible in the [diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png)  should be defined and set up in the constructor method.
- [x]  User interface : console app entry-point is created
- [x]  Set up associations between classes and objects.
- [x]  All data should be preserved by saving collections in .json files.
- [x] Build interactive console apps
- [x] Create a schema.sql file with tables that will be analogical to the structure of the classes in your app

### CODE REVIEW

File: spec/author_spec.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        new_author = Author.new('first name', 'Last name')
+        item1.add_author(new_author)
+        expect(new_author.items).to include item1
+
+        item1.add_author(new_author)
+        puts new_author.items
+        expect(new_author.items.length).to eq 1
+      end
+    end
+  end
```
Comment (danmainah) [inline spec/author_spec.rb:41]: - [ ] @Team great work adding tests for the author class however you forgot to add tests for these methods: to_s,  to_json, and self.from_json
image

File: spec/genre_spec.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    end
+  end
+  context 'When testing genre methods' do
+    it 'Should add a new item to the items array' do
+      new_genre = Genre.new('Test')
+      item = Item.new(2)
+      new_genre.add_item(item)
+      expect(new_genre.items.length).to eq 1
+    end
+  end
```
Comment (danmainah) [inline spec/genre_spec.rb:19]: - [ ] @Team great work adding tests for the genre class however you forgot to add tests for these methods: to_s,to_json, and self.from_json

Comment (danmainah) [reply spec/genre_spec.rb:19]: - [ ] @Team I noted you have not added tests for the item class kindly consider adding tests for this class

File: spec/book_spec.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+      expect(@book2.archived).to eq true
+    end
+    it 'Book was published >= 10 yrs ago' do
+      # when the book is >= 10 yrs old it can be archived!
+      @book10.move_to_archive
+
+      expect(@book10.archived).to eq true
+    end
+  end
```
Comment (danmainah) [inline spec/book_spec.rb:37]: - [ ] @Team great work adding tests for the book class however you forgot to add tests for these methods: to_s,to_json, and self.from_json

File: spec/album_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      new_album = MusicAlbum.new(true, '20210101')
+      expect(new_album.to_json(new_album)).to eq '{"on_spotify":true,"published_date":"2021-01-01","archived":false}'
+    end
+    it 'Should create instance from json' do
+      # rubocop:disable Layout/LineLength
+      new_album = MusicAlbum.from_json(JSON.parse('{"on_spotify":false,"published_date":"2021-01-01","archived":false}'))
+      # rubocop:enable Layout/LineLength
+      expect(new_album.on_spotify).to eq false
+    end
+  end
```
Comment (danmainah) [inline spec/album_spec.rb:40]: - [ ] @Team great work adding tests for the music album class however you forgot to add tests for these methods: to_s,to_json, and self.from_json

File: spec/game_spec.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        new_game2.move_to_archive
+        expect(new_game2.archived).to eq true
+
+        # with an incorrect date today date will be added to the item class
+        new_game3 = Game.new('test', true, 1, 15)
+        new_game3.move_to_archive
+        expect(new_game3.archived).to eq false
+      end
+    end
+  end
```
Comment (danmainah) [inline spec/game_spec.rb:32]: - [ ] @Team great work adding tests for the game class however you forgot to add tests for these methods: to_s,to_json, and self.from_json

File: sql/schema.sql [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  source varchar(30),
+  label varchar(30),
+  publish_date int,
+  archived boolean,
+  author_id int,
+  game_id int,
+  album_id int, 
+  genre_id int,
+  CONSTRAINT kf_albums FOREIGN KEY(album_id) REFERENCES albums(id) ON DELETE CASCADE,
+  CONSTRAINT kf_genres FOREIGN KEY(genre_id) REFERENCES genres(id) ON DELETE CASCADE
```
Comment (danmainah) [inline sql/schema.sql:53]: - [ ] @Team you forgot to add references for the game_id and author_id ,kindly add their references.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Benawi , @kayjeee and @derrick1451 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️  Ensured all the options in the UI are working correctly
✔️  Ensured data is stored correctly after the user exits the app.
✔️  All the tests added are passing.


# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Comment by Benawi** [comment]
## Changes ♻
- [x] A new video recorded [👁 Walkthrough](https://drive.google.com/file/d/1p-FD-SK1DSDMRLGiTpuNjYJOdMWLcB-T/view?usp=sharing)

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @Benawi, @kayjeee and @derrick1451 👏 👏 👏 👏, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

so-close

## Highlights
Engaging presentation within time frame ✔️ 
Implemented tests for `to.s and to_json` methods ✔️ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Benawi** [comment]
## Changes ♻
- [x] All the repeated tables (Items) removed from [sql/schema.sql](https://github.com/derrick1451/ruby-capstone/pull/56/files/04031d00043c640e31713e814e8c87c44acd5428#diff-c4c0021442fe03b6f1aebb62fed2fe167c0016862c2f74d2689f141a010ea9ad)

**Review by KanzaTahreem — APPROVED** [review]
# Approved 🙌

Hi @team

Good job implementing the suggested changes 💪 **_Your project is COMPLETE! There is nothing else to say other than... it's time to merge it! 🚀 Congratulations!_**



## Highlights 🎯
There are some excellent points that you have made perfectly. Keep up the great work! 👏

-  ✅ Repeated tables are removed
-  ✅ Your README looks good
-  ✅ Linters are working properly

## Optional suggestions ❕
N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👨‍💻💻

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**